File: lumen-capabilities/prices/precio_bs.py
#!/usr/bin/env python3
"""
Capability: precio_bs
Devuelve la tasa USD/VES paralelo y oficial desde ve.dolarapi.com.
Llamado por la skill `tropico-prices` del kit Tropico Wallet.

Uso:
    python3 precio_bs.py --instance tropico-mvp [--pretty]

Output: JSON estructurado con `usdToBs`, `usdToBsOficial`, `fuente`, `fetchedAt`.
"""

import argparse
import json
import sys
from datetime import datetime, timezone
from urllib.request import urlopen, Request
from urllib.error import URLError

DOLARAPI_URL = "https://ve.dolarapi.com/v1/dolares"
TIMEOUT_SECONDS = 5


def fetch_precio_bs() -> dict:
    """Fetch tasa from ve.dolarapi.com. Returns dict with usdToBs (paralelo), usdToBsOficial (BCV)."""
    try:
        req = Request(DOLARAPI_URL, headers={"User-Agent": "tropico-lumen/0.1"})
        with urlopen(req, timeout=TIMEOUT_SECONDS) as response:
            data = json.loads(response.read().decode("utf-8"))
    except URLError as e:
        return {
            "ok": False,
            "error": f"Network error: {e.reason}",
            "fuente": "fallback",
        }
    except Exception as e:
        return {
            "ok": False,
            "error": f"Error fetching precio: {str(e)}",
            "fuente": "fallback",
        }

    paralelo = next((d for d in data if d.get("fuente") == "paralelo"), None)
    oficial = next((d for d in data if d.get("fuente") == "oficial"), None)

    paralelo_rate = paralelo and (paralelo.get("promedio") or paralelo.get("venta") or paralelo.get("compra"))
    oficial_rate = oficial and (oficial.get("promedio") or oficial.get("venta") or oficial.get("compra"))

    if not paralelo_rate or paralelo_rate <= 0:
        return {
            "ok": False,
            "error": "Invalid paralelo rate from API",
            "fuente": "fallback",
        }

    return {
        "ok": True,
        "usdToBs": paralelo_rate,
        "usdToBsOficial": oficial_rate,
        "fuente": "ve.dolarapi.com",
        "fetchedAt": datetime.now(timezone.utc).isoformat(),
    }


def main():
    parser = argparse.ArgumentParser(description="Get USD/VES rate from ve.dolarapi.com")
    parser.add_argument("--instance", required=True, help="Tropico instance ID")
    parser.add_argument("--pretty", action="store_true", help="Pretty print JSON")
    args = parser.parse_args()

    result = fetch_precio_bs()

    if args.pretty:
        print(json.dumps(result, indent=2, ensure_ascii=False))
    else:
        print(json.dumps(result, ensure_ascii=False))

    sys.exit(0 if result.get("ok") else 1)


if __name__ == "__main__":
    main()

File: lib/precio-bs.ts
/**
 * Fetch + cache de la tasa USD/VES.
 *
 * Fuente: ve.dolarapi.com — comunidad, devuelve oficial (BCV) y paralelo.
 * Tropico usa el paralelo porque es la realidad para el venezolano que ahorra.
 *
 * Esto se llama desde el API route /api/precio-bs (no exponer cache server-only al client).
 */

const DOLARAPI_URL = "https://ve.dolarapi.com/v1/dolares";

export type PrecioBs = {
  /** Tasa paralelo (la realidad VE) */
  usdToBs: number;
  /** Tasa BCV oficial (referencia) */
  usdToBsOficial: number | null;
  fuente: "ve.dolarapi.com" | "fallback";
  fetchedAt: number; // epoch ms
};

const FALLBACK: PrecioBs = {
  usdToBs: 650, // último valor conocido — actualizar manualmente si hace falta
  usdToBsOficial: 500,
  fuente: "fallback",
  fetchedAt: 0,
};

let cache: PrecioBs | null = null;
const CACHE_MS = 60_000; // 1 min — la tasa no cambia tan rápido como para refrescar cada 30s

type DolarApiEntry = {
  moneda: string;
  fuente: "oficial" | "paralelo";
  nombre: string;
  promedio: number | null;
  compra: number | null;
  venta: number | null;
  fechaActualizacion: string;
};

export async function fetchPrecioBs(): Promise<PrecioBs> {
  if (cache && Date.now() - cache.fetchedAt < CACHE_MS) {
    return cache;
  }

  try {
    const res = await fetch(DOLARAPI_URL, {
      next: { revalidate: 60 },
      signal: AbortSignal.timeout(5000),
    });
    if (!res.ok) throw new Error(`status ${res.status}`);

    const data = (await res.json()) as DolarApiEntry[];
    const paralelo = data.find((d) => d.fuente === "paralelo");
    const oficial = data.find((d) => d.fuente === "oficial");

    const paraleloRate = paralelo?.promedio ?? paralelo?.venta ?? paralelo?.compra;
    const oficialRate = oficial?.promedio ?? oficial?.venta ?? oficial?.compra;

    if (typeof paraleloRate !== "number" || paraleloRate <= 0) {
      throw new Error("invalid paralelo rate");
    }

    cache = {
      usdToBs: paraleloRate,
      usdToBsOficial: typeof oficialRate === "number" ? oficialRate : null,
      fuente: "ve.dolarapi.com",
      fetchedAt: Date.now(),
    };
    return cache;
  } catch {
    if (cache) return cache;
    return FALLBACK;
  }
}

export function usdToBs(usd: number, rate: PrecioBs): number {
  if (!Number.isFinite(usd)) return 0;
  return usd * rate.usdToBs;
}

export function bsToUsd(bs: number, rate: PrecioBs): number {
  if (!Number.isFinite(bs) || rate.usdToBs <= 0) return 0;
  return bs / rate.usdToBs;
}
